validator/client/beacon-api/doppelganger.go

```diff
@@ -0,0 +1,229 @@
+package beacon_api
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"strconv"
+
+	"github.com/pkg/errors"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v3/runtime/version"
+	"github.com/prysmaticlabs/prysm/v3/time/slots"
+)
+
+type DoppelGangerHelper struct {
+	validatorEpoch types.Epoch
+	response       *ethpb.DoppelGangerResponse_ValidatorResponse
+}
+
+func (c *beaconApiValidatorClient) checkDoppelGanger(ctx context.Context, in *ethpb.DoppelGangerRequest) (*ethpb.DoppelGangerResponse, error) {
+	// Check if there is any doppelganger validator for the last 2 epochs.
+	// - If we are in Phase0, we consider there is no doppelganger.
+	// - If all validators we want to check doppelganger existence were live in local antislashing
+	//   database for the last 2 epochs, we consider there is no doppelganger.
+	//   This is typically the case when we reboot the validator client.
+	// - If some validators we want to check doppelganger existence were NOT live
+	//   in local antislashing for the last two epochs, then we check onchain if there is
+	//   some liveness for these validators. If yes, we consider there is a doppelganger.
+
+	// Check inputs are correct
+	if in == nil || in.ValidatorRequests == nil || len(in.ValidatorRequests) == 0 {
+		return &ethpb.DoppelGangerResponse{
+			Responses: []*ethpb.DoppelGangerResponse_ValidatorResponse{},
+		}, nil
+	}
+
+	validatorRequests := in.ValidatorRequests
+
+	// Prepare response
+	stringPubKeys := make([]string, len(validatorRequests))
+	stringPubKeyToDoppelGangerHelper := make(map[string]DoppelGangerHelper, len(validatorRequests))
```

Comment:
Rename variable to `stringPubKeyToDoppelGangerInfo` if you decide to rename `DoppelGangerHelper` to `DoppelGangerInfo`